typechecker+codegen: Emit new bodies for bindings with different types (#1602)
SerenityOS/jakt: documentation/pattern-match-avoid-duplication.md

the matched value may conform to.
- `binding1` and `binding2` are "bindings". They are variables that extract
different parts of the structure that the value is checked to conform to. We
can then use those variables inside `<body1>`.
- `body1` and `body2` are "bodies". They are the pieces of code that we want to
execute on certain cases.
- `pat1` and `pat2` form a "pattern group". Matching either of them will branch
to the code specified in `<body1>`.
- Pattern groups and bodies together form a "case". Each case has a single body
and a group of patterns.

## The problem

### Constraint 1: Multiple binding types

Jakt allows us to use bindings of different types in the same pattern group, as
long as the body typechecks against both "versions" of the bindings. For an
illustration, take this valid Jakt sample:

```jakt
enum TypedPair {
    I32(a: i32, b: i32)
    U64(a: u64, b: u64)
}

fn get_pair() -> TypedPair { ... }

fn main() {
    match get_pair() {
        I32(a, b) | U64(a, b) => {
            let result = a + b
            eprintln("result: {}", result)
        }
    }
}
```

Notice that the first body of the match accepts both types of `a` and `b`
(`i32` and `u64`):

> - `a + b` typechecks as long as `a` implements `Add` with `b` as its RHS.

Reviewer: ```suggestion
- `a + b` typechecks as long as the type of `a` implements `Add` with `b` as its RHS.
```

- `eprintln` accepts both result types.

According to the rule above, this sample should typecheck and always print the
added result.

We want to generate the minimal amount of code that contains both versions of
the branching, meaning that paths for both `I32` and `U64` should get their
`result` to be `i32` and `u64` respectively.

When generating the code for that to work, even though we're lenient on the
user, we still have to separate the code path that uses `u64` from the code
path that uses `i32`. Since we generate C++ code, we can rely on C++'s
substitution power for things like calls to overloaded functions, but wherever
we have to fully specify the type of `a`, `b` or `result` we will have to
generate two versions of the bodies.

## A previous solution: Why is it slow?

The first implementation ran through all of the cases, typechecking the body
for each of the patterns and submitting a new case per pattern, with their
respective body.

When generating code for a `match`, codegen sees a `CheckedExpression::Match`

> containing a list of `CheckedMatchCase`s. Each `CheckedMatchCase` exactly one

Reviewer: ```suggestion
containing a list of `CheckedMatchCase`s. Each `CheckedMatchCase` contains exactly one
```

pattern and one body that uses the bindings declared within said pattern.
 
The problem with this solution is that it effectively duplicates the body as
many times as patterns exist in the same group. This made compiling painfully
slow, but the overhead on typechecking the body was not the main issue: Codegen
would receive the bodies already duplicated, severely limiting what can be done
at late codegen stage. Take into account that the codegen stage has the most
complete version of the program that Jakt can have, since it needs all the
concrete types that are used in order to generate working C++. In #1597 we can
see that a ~300 line Jakt function (`typecheck_binary_operation`) is compiled
to ~5500 lines of C++ code! That happened because the bodies for each case are
large, and the bodies were duplicated many times.

To ensure that similar functions to `typecheck_binary_operation` do not make

> the C++ compiler sit through 18x the amount of Jakt code, we'll add a second

Reviewer: ```suggestion
the C++ compiler sift through 18x the amount of Jakt code, we'll add a second
```
?

        }
    }
}
```

#### Constraint 2.2: Multiple patterns with no bindings
```jakt
enum Answer {
    Is42
    IsNot42(actual_value: u32)
    IsNot42AndNoValue
}

fn get_answer() -> Answer { ... }

fn main() {
    match get_answer() {
        Is42 => {
            // long body 1
        }
        IsNot42 | IsNot42AndNoValue => {
            // long body 2
        }
    }
}
```

In both cases, we want Jakt to ideally only generate the long bodies once.

## A partial improvement (#1597)

PR #1597 addressed Constraint 2.1 by keeping the shared relationship between
the patterns inside a pattern group and the body they jump to. This enables
codegen to find pattern groups that have no bindings on them and easily group
them by concatenating `case` labels or concatenating boolean expressions with
`||`.

Now each `CheckedMatchCase` contains a list of `CheckedMatchPattern` and one
body that uses the bindings declared within said pattern. The implementation
typechecks all bodies like the last version, but only selects the last checked
body.

The problem with this approach is that now the typechecker portion of this
solution does not respect Constraint 1, since the last checked body is not

> enough to represent the case if bindings from the same parsed pattern group can

Reviewer: ```suggestion
enough to represent the case if bindings from the same parsed pattern group
```

have different binding types.

> This makes `tests/codegen/match_enum_same_body_different_impl.jakt` to fail
> codegen, since it calls `maybe_templated<JaktArray<i32>>` on all cases, when

Reviewer: ```suggestion
This causes `tests/codegen/match_enum_same_body_different_impl.jakt` to fail,
since it codegens a call to `maybe_templated<JaktArray<i32>>` on all cases, when
```

Each `CheckedMatchCase` contains a list of `CheckedMatchPattern`, a set of
bindings `[String:VarId]` and a body where those bindings are used. All the
patterns in this `CheckedMatchCase` are known to have the same set of bindings
as the case.

### How it works

Typechecker walks the patterns to get the type information out of them as it
did before, but instead of adding the binding variables to a scope, it adds
them to a `BindingKeyBuilder`. Typechecker "submits" bindings to the builder,
which then produces a `BindingKey` that tells us whether the set of bindings
that were declared already exist within the same pattern group.

Since we don't want to be passing a lot of state regarding the match to helper
functions like `.typecheck_match_variant`, `BindingKeyBuilder` only contains a
reference to the slice of cases that have been declared and refer to the same
`ParsedMatchBody`. Remember that it could be multiple due to Constraint 1.

To enforce that the set of cases checked by `BindingKeyBuilder` are the correct
set, and that context is not lost when reading through the long loop bodies of
`.typecheck_match`, all the state regarding match typecheck is in

> `MatchBuilder`. This struct has two functionalities: - It will make us

Reviewer: where second functionality

Author: oh, I incorrectly removed the bullet point for the second paragraph.
